packages/vm/lib/runTx.ts

```diff
@@ -63,6 +66,14 @@ export interface RunTxOpts {
    * the `generateAccessList()` method must be implemented.
    */
   reportAccessList?: boolean
+
+  /**
+   * Optional clique Address: if the consensus algorithm is on clique,
+   * and this parameter is provided, use this as the beneficiary of transaction fees
+   * If it is not provided and the consensus algorithm is clique, instead
+   * get it from the block using `cliqueSigner()`
+   */
+  cliqueBeneficiary?: Address
```

Comment:
Would it make sense to expand this option and allow to pass in a miner of clique beneficiary independent from the consensus type?

Reply:
Ah, this is still an interesting/important question it would be good to get additional opinions on and not just let this slip through. //cc @jochem-brouwer @ryanio

Reply:
I am not super sure what you mean by this. The clique beneficiary is actually not present in the header of the block (miner is always set to the zero address IIRC). To get the actual clique beneficiary, you have to get the address which signed the extraData of the block.

Reply:
I mean if we should allow here to generally add the `minerOrBeneficiary` as this option and not only for the clique case.

This option is used in the current form:

```typescript
if (this._common.consensusType() === 'pow') {
    miner = block.header.coinbase
  } else {
    // Backwards-compatibilty check
    // TODO: can be removed along VM v5 release
    if (opts.cliqueBeneficiary) {
      miner = opts.cliqueBeneficiary
    } else {
      if ('cliqueSigner' in block.header) {
        miner = block.header.cliqueSigner()
      } else {
        miner = Address.zero()
      }
    }
  }
```

So it also might make sense to use this in the PoW case and if present than set the `miner` to the miner from this option and not to the eventual default value from `block.header.coinbase`.

Does this makes sense? I am not saying that we *should* do this, just saying that this might be a worthwhile generalization of the option which we should at least consider.

Reply:
hey, catching up on this thread now.

ok so I see this problem was introduced when I moved the tx receipt generation to runTx, where we need a block with the current accurate gasUsed to generate the right receipt.

from @jochem-brouwer ([comment link](https://github.com/ethereumjs/ethereumjs-monorepo/pull/1148#issuecomment-827139618)):
> Ok, this commit 8c112f0 fixes the consensus bug at block 55. The problem had nothing to do with EIP1559, it had to do with runBlock/runTx in VM: runBlock originally fed runTx an edited block than the block which is actually being ran; the gasUsed field is changed and used to keep track of the cumulative gas used. This commit is a hotfix and might be changed in the future (probably removing editing the block and just passing along the block which is actually being ran in the context).
> 
> We should add a test to ensure that clique blocks with txs pay the right miner, which is not tested in VM currently.

I can foresee future problems with this "modified block" being passed in and think it would be good to go back to the original block without any fancy meddling. Thinking about it with a fresh perspective now, what do you guys think if I add a `blockGasUsed` parameter to runTx instead, expressly for accurate tx receipt generation? Then we don't have modify the block for this reason.

I'm also confused which part of the code this todo references now:
`// TODO: can be removed along VM v5 release`
could you help clarify @holgerd77? I can add to the PR I would like to open for the work above.

Reply:
> Thinking about it with a fresh perspective now, what do you guys think if I add a blockGasUsed parameter to runTx instead, expressly for accurate tx receipt generation? Then we don't have modify the block for this reason.

I think that's a good idea.

The TODO was for `if ('cliqueSigner' in block.header)`, this was the discussion that this would break if used with a `Block` library which hasn't included the clique changes yet. (I think v5 might be a mistake and this should be v6?)

> In conclusion, if my suggestion above is acceptable, we can remove this newly introduced cliqueBeneficiary option in runTx in favor of a blockGasUsed parameter and i can add some tests to ensure the beneficiary is correctly derived.

Yeah, sounds great, leaves me somewhat relieved.

😄